### PR TITLE
connlib: refactoring tunnel for performance and readablity

### PR DESCRIPTION
**Pull request by conectado**
Splits of from #2104 

Some refactoring I did for firezone/product#681 to obtain better flamegraphs, additionally to improve performance I added a `wire` target so we can turn off wire tracing in production and generally tried to improve logs for easier filtering and use of structured-tracing.

### CODE REVIEW

File: rust/connlib/libs/tunnel/src/control_protocol.rs [hunk 1]
```diff
@@ -145,12 +146,11 @@ where
         gateway_id: Id,
         resource_id: Id,
     ) {
-        tracing::trace!("Peer Connection State has changed: {state}");
+        tracing::trace!(?state, "peer_state");
```
Comment (thomaseizinger) [inline rust/connlib/libs/tunnel/src/control_protocol.rs]: The state should already be in the `span` right? Seems redundant to log it here too.

File: rust/connlib/libs/tunnel/src/control_protocol.rs [hunk 2]
```diff
@@ -345,7 +352,7 @@ where
         let p_key = preshared_key.clone();
         data_channel.on_open(Box::new(move || {
             Box::pin(async move {
-                tracing::trace!("new data channel opened!");
+                tracing::trace!("data_channel");
```
Comment (thomaseizinger) [inline rust/connlib/libs/tunnel/src/control_protocol.rs]: To be frank, I liked the old message better (without punctuation though).

File: rust/connlib/libs/tunnel/src/device_channel_unix.rs [hunk 3]
```diff
@@ -1,74 +1,78 @@
...
             let mut guard = self.0.readable().await?;
 
-            match guard.try_io(|inner| {
-                inner.get_ref().read(out).map_err(|err| match err {
-                    Error::IfaceRead(e) => e,
-                    _ => panic!("Unexpected error while trying to read network interface"),
-                })
-            }) {
-                Ok(result) => break result.map(|e| e.len()),
+            match guard.try_io(|inner| inner.get_ref().read(out)) {
```
Comment (thomaseizinger) [inline rust/connlib/libs/tunnel/src/device_channel_unix.rs]: In recent versions of tokio, you can use `async_io`: https://docs.rs/tokio/latest/tokio/io/unix/struct.AsyncFd.html#method.async_io

File: rust/connlib/libs/tunnel/src/peer.rs [hunk 4]
```diff
@@ -181,4 +191,52 @@ impl Peer {
...
         self.translated_resource_addresses.write().insert(addr, id);
     }
+
+    pub(crate) fn encapsulate<'a>(
+        &self,
+        src: &'a mut [u8],
+        dst: &'a mut [u8],
+    ) -> Result<EncapsulatedPacket<'a>> {
+        let Some(mut packet) = MutableIpPacket::new(src) else {
+            tracing::error!("Developer error: we should never see a packet through the tunnel wire that isn't ip");
```
Comment (thomaseizinger) [inline rust/connlib/libs/tunnel/src/peer.rs:202]: Should we make these `debug_assert`s?

**Review by thomaseizinger — APPROVED** [review]
I am not very familiar with most of the code so I only checked it on a high-level :)

File: rust/connlib/libs/common/src/error.rs [hunk 5]
```diff
@@ -108,6 +108,15 @@ pub enum ConnlibError {
     /// Invalid phoenix channel reference
     #[error("Invalid phoenix channel reply reference")]
     InvalidReference,
+    /// Invalid packet format
+    #[error("Received baddly formatted packet")]
```
Comment (jamilbk) [inline rust/connlib/libs/common/src/error.rs]:
```suggestion
    #[error("Received badly formatted packet")]
```

**Review by jamilbk — APPROVED** [review]
Just a minor typo

**Review by thomaseizinger — APPROVED** [review]

**Pull request merged**